MapboxCoreNavigation/Offline.swift

```diff
@@ -0,0 +1,122 @@
+    static var translationsPath: URL {
+        let path = coreNavigationCacheDirectory.appendingPathComponent("translations")
+        ensureDirectoryExist(at: path)
+        return path
+    }
+    
+    static var coreNavigationCacheDirectory: URL {
+        let cacheDirectory = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)[0]
+        return URL(fileURLWithPath: cacheDirectory).appendingPathComponent(Bundle.mapboxCoreNavigation.bundleIdentifier!).appendingPathComponent("offline")
+    }
+    
+    // Ensures that a directory exist at the given path.
+    // Returns false if no directory was created.
+    @discardableResult
+    static func ensureDirectoryExist(at path: URL) -> Bool {
+        do {
+            try FileManager.default.createDirectory(at: path, withIntermediateDirectories: true, attributes: nil)
+            return true
+        } catch {
+            return false
+        }
+    }
+}
+
+public typealias RouteCompletionHandler = (_ waypoints: [Waypoint]?, _ routes: [Route]?, _ error: NSError?) -> Void
+
+public class ExtendedDirections: NSObject {
+    
+    public static let shared = ExtendedDirections()
+    
+    static let offlineSerialQueueLabel = Bundle.mapboxCoreNavigation.bundleIdentifier!.appending(".offline")
+    
+    static let serialQueue = DispatchQueue(label: ExtendedDirections.offlineSerialQueueLabel)
+    
+    var _navigator: MBNavigator!
+    var navigator: MBNavigator {
+        //assert(OperationQueue.current?.underlyingQueue?.label == ExtendedDirections.offlineSerialQueueLabel,
+        //       "The offline navigator must be accessed from the dedicated serial queue")
+        if _navigator == nil {
+            self._navigator = MBNavigator()
+            
+            let tilePath = OfflineStorage.tilesPath.absoluteString.replacingOccurrences(of: "file://", with: "")
+            let localizationPath = OfflineStorage.translationsPath.absoluteString.replacingOccurrences(of: "file://", with: "")
+            
+            // navigator raises an exception after each try:
+            // 1: uncompressed tiles packed in a tar archive,
+            // 2: uncompressed tiles in directories
+            // 3: compressed tiles in directories
+            // We are using the 3rd option
+            self._navigator.configureRouter(forTilesPath: tilePath, translationsPath: localizationPath)
+        }
+        
+        return _navigator
+    }
+    
+    open func calculate(_ options: RouteOptions, completionHandler: @escaping RouteCompletionHandler) {
+        
+        guard let navigationOptions = options as? NavigationRouteOptions else {
+            Directions.shared.calculate(options, completionHandler: completionHandler)
+            return
+        }
+        
+        switch navigationOptions.preferredRequestOption {
+        case .preferClientSide:
+            fatalError("Not yet implemented")
+        case .preferServerSide:
+            fatalError("Not yet implemented")
```

Comment:
what's the thinking here - preflighting an HTTP request is [discouraged/considered an anti-pattern](https://developer.apple.com/library/archive/documentation/NetworkingInternetWeb/Conceptual/NetworkingOverview/WhyNetworkingIsHard/WhyNetworkingIsHard.html) these days